File: src/02.string/04.transform/11.8.ts
export function myAtoi(s: string): number {
    let res = 0;
    let sign = 0;
    let numIndex = 0;
    let isNum = false;
    let hasNum = (c: string): boolean => {
        return c >= "0" && c <= "9";
    };

    for (let i = 0; i < s.length; i++) {
        if (!isNum) {
            if (s[i] === " ") {
                continue;
            } else if (hasNum(s[i]) || s[i] === "-" || s[i] === "+") {
                if (s[i] === "-" || s[i] === "+") {
                    sign = s[i] === "-" ? -1 : 1;
                } else {
                    res = s[i].charCodeAt(0) - 48;
                    if (sign === 0) sign = 1;
                }
                isNum = true;
                numIndex = i + 1;
            } else {
                break;
            }
        } else if (hasNum(s[i]) && numIndex === i) {
            res *= 10;
            res += s[i].charCodeAt(0) - 48;
            numIndex = i + 1;
        } else {
            break;
        }
    }

    res *= sign;

    if (res > 2147483647) {
        res = 2147483647;
    } else if (res < -2147483648) {
        res = -2147483648;
    }

    return res;
}

File: src/02.string/04.transform/11.8.go
package transform

import "math"

func MyAtoi(s string) int {
	var res int64 = 0
	isNum, numIndex, sign := false, 0, 0
	hasNum := func(c rune) bool {
		return c >= '0' && c <= '9'
	}

	for i, v := range s {
		if !isNum {
			if v == ' ' {
				continue
			} else if v == '-' || v == '+' {
				numIndex = i + 1
				isNum = true
				if v == '-' {
					sign = -1
				} else {
					sign = 1
				}
			} else if hasNum(v) {
				res = int64(v - '0')
				numIndex = i + 1
				isNum = true
				if sign == 0 {
					sign = 1
				}
			} else {
				break
			}
		} else if hasNum(v) && numIndex == i {
			res *= 10
			res += int64(v - '0')
			numIndex = i + 1
			if res > math.MaxInt32+1 {
				res = math.MaxInt32 + 1
				break
			}
		} else {
			break
		}
	}

	res *= int64(sign)
	if res > math.MaxInt32 {
		res = math.MaxInt32
	} else if res < math.MinInt32 {
		res = math.MinInt32
	}

	return int(res)
}
